docs/code-guide.md

```diff
@@ -0,0 +1,133 @@
+- Try to flatten the UI state into a basic calculation, deriving data from props
+- Create a new component abstraction when you're nesting conditional logic or top-level if/else statements. Ternaries are reserved for small, easily readable logic.
+- When complex data manipulation and logic is necessary, make use of custom hooks.
+- Avoid passing whole objects to components when they only need a few properties. It will help clarify which information they rely on.
+- Presentational components should not contain any domain/business logic.
+- Avoid putting dependent logic inside `useEffect`; it causes misdirection of what the logic is doing. Choose to explicitly define logic rather than depend on implicit reactive behavior
+- Prefer state machines over multiple related `useState` calls. Multiple interdependent state variables make code harder to reason about.
+- Avoid `setTimeouts`. They are flaky and usually a _hack_, always provide a comment on _why_ you are using them. This doesn't affect if the "code runs" or not most of the time, but they can introduce subtle bugs that can grow into big issues that aren't obvious until someone goes in and has to spend a lot of time refactoring everything.
+
+## React Hooks Patterns
+
+- Avoid using `useMemo` for processes that are not computationally expensive
+- Functions are memoized with `useCallback` only when necessary:
+  * Functions used as dependencies in `useEffect`, `useMemo`, or other hooks
+  * Functions passed to context providers or returned from custom hooks
+  * Functions passed as props to memoized child components (to prevent unnecessary re-renders)
+- Functions are NOT unnecessarily memoized:
+  * Simple event handlers that aren't passed as props
+  * Functions without dependencies or only used within the same component
+- All hook dependencies (`useEffect`, `useMemo`, `useCallback`) use referentially stable variables
+- No `useEffect`s that take incoming props and compute them for a local `useState`, this is `useMemo` with extra steps
+- Make sure the [exhaustive-deps rule](https://react.dev/reference/eslint-plugin-react-hooks/lints/exhaustive-deps) is applied. If you encounter a file where the ESLint rule is ignored (`// eslint-disable-next-line react-hooks/exhaustive-deps`), re-enable it and fix the dependency array. If for any reason the rule has to be skipped, provide a comment explaining why.
+
+### More On `useEffect`
+
+Make sure to follow React guidelines on `useEffect`: [You Might Not Need an Effect](https://react.dev/learn/you-might-not-need-an-effect)
+
+Common cases where `useEffect` is NOT needed:
+
+* Transforming data for rendering (use variables or useMemo instead)
+* Handling user events (use event handlers instead)
+* Resetting state when props change (use key prop or calculate during render)
+* Updating state based on props/state changes (calculate during render)
+* Set default values for input fields
```

Comment:
is it safe to maybe just say _'set values for input fields'_?

like any value, not only default..

(examples abound - here's one:  https://github.com/RedHatInsights/uhc-portal/blob/4a3a9f7ea11103976035f6d45826b9e943add512/src/components/clusters/wizards/osd/BillingModel/BillingModel.tsx#L147)

Reply:
@eliranmal isn't your example falling into the category "Updating state based on props/state changes"?